linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,513 @@
+{
+    if (param == NULL)
+    {
+        LogCritical("Invalid args: param: %p", param);
+    }
+    else
+    {
+        struct timespec ts;
+        THREADPOOL* threadpool = param;
+        int64_t current_index;
+
+        do
+        {
+            if (clock_gettime(CLOCK_REALTIME, &ts) == -1)
+            {
+                // Consider Log
+                LogError("Failure getting time from clock_gettime");
+            }
+            else
+            {
+                // Setup the timeout for the semaphore to 2 sec
+                ts.tv_sec += 2;
+                if (sem_timedwait(&threadpool->semaphore, &ts) != 0)
+                {
+                    // Timed out
+                }
+                else
+                {
+                    THREADPOOL_WORK_FUNCTION task_func = NULL;
+                    void* task_param = NULL;
+                    // Is this index waiting for us to complete it
+                    srw_lock_acquire_shared(threadpool->srw_lock);
+                    {
+                        int32_t existing_count = interlocked_add(&threadpool->task_array_size, 0);
+                        // We have a new item
+                        current_index = interlocked_increment_64(&threadpool->consume_idx) % existing_count;
+
+                        TASK_RESULT curr_task_result = interlocked_compare_exchange(&threadpool->task_array[current_index].task_state, TASK_WORKING, TASK_WAITING);
+                        if (TASK_WAITING == curr_task_result)
+                        {
+                            task_func = threadpool->task_array[current_index].task_func;
+                            task_param = threadpool->task_array[current_index].task_param;
+                            (void)interlocked_exchange(&threadpool->task_array[current_index].task_state, TASK_NOT_USED);
+                        }
+                        else
+                        {
+                            // Do nothing here, this should not happen, but log if it does
+                            LogWarning("The impossible has happend, someone got the item %" PRId64 " task value %" PRI_MU_ENUM "", current_index, MU_ENUM_VALUE(TASK_RESULT, curr_task_result));
+                        }
+                    }
+                    srw_lock_release_shared(threadpool->srw_lock);
+
+                    if (task_func != NULL)
+                    {
+                        task_func(task_param);
+                    }
+                }
+            }
+
+        } while (interlocked_add(&threadpool->state, 0) != THREADPOOL_STATE_CLOSING);
+    }
+    return 0;
+}
+
+static int reallocate_threadpool_array(THREADPOOL* threadpool)
+{
+    int result;
+    {
+        // Lock the array here
+        srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+        int32_t existing_count = interlocked_add(&threadpool->task_array_size, 0);
+
+        // Double the items
+        (void)interlocked_exchange(&threadpool->task_array_size, existing_count*2);
```

Comment:
>existing_count*2 [](http://example.com/codeflow?start=65&length=16)

unlikely to have this many items but we should probably check for overflow. I think we are just 16 doublings away from UINT32_MAX and we should probably fail and log if that were to somehow happen.